### PR TITLE
add preferred_serialization_format and msgpack support

### PR DESCRIPTION
**Pull request by bastelfreak**
more information here https://docs.puppetlabs.com/puppet/latest/reference/experiments_msgpack.html#docs-»-experimental-features:-msgpack-support

add general support for preferred_serialization_format in master and agent, also install msgpack if needed

### CODE REVIEW

File: manifests/agent.pp [hunk 1]
```diff
@@ -310,15 +300,35 @@
...
     value   => $configtimeout,
   }
   if $stringify_facts != undef {
     ini_setting {'puppetagentstringifyfacts':
-      ensure  => present,
       setting => 'stringify_facts',
       value   => $stringify_facts,
     }
   }
+  if $serialization_format != undef {
```
Comment (stephenrjohnson) [inline manifests/agent.pp]: Are these already packaged for use

libmsgpack-ruby1.8 - binary-based efficient object serialization library for ruby1.8
libmsgpack-ruby1.9.1 - binary-based efficient object serialization library for ruby1.9.1

**Comment by bastelfreak** [comment]
these are debian packages, I'm on redhat and I can't find suitable rpms. if you wan't I can make a switch for installing these packages or the gems. If a package name is provided, it is installed, otherwise the gem?

**Comment by bastelfreak** [comment]
@stephenrjohnson could you again take a look? now it is also possible to install packages

File: manifests/master.pp [hunk 2]
```diff
@@ -313,6 +315,27 @@
       setting => 'digest_algorithm',
       value   => $digest_algorithm,
   }
-
+  if $serialization_format != undef {
```
Comment (stephenrjohnson) [inline manifests/master.pp]: You still need to add the option in install the package here as well.

**Comment by bastelfreak** [comment]
if a packagename  is set, it gets installed in the lines 315-319?

**Comment by stephenrjohnson** [comment]
No if you install msgpack on the master you use the gem version and build it from source you don't install the rpm or deb packaged version.

**Comment by bastelfreak** [comment]
ups. I see, I just did it for the agent. Will fix it later.

**Comment by bastelfreak** [comment]
Just tested it heavily with >2100 nodes and 3 masters. I added a few additional "unless defined" statements so this won't interference with other modules. In my Opinion it is now ready for merge.

**Comment by bastelfreak** [comment]
@stephenrjohnson could you have a look again?

File: manifests/master.pp [hunk 3]
```diff
@@ -313,6 +315,36 @@
...
+    if $serialization_package != undef {
```
Comment (pall-valmundsson) [inline manifests/master.pp]: Isn't this missing as a parameter for the `master` class?

Comment (stephenrjohnson) [inline manifests/master.pp]: Fixed in last commit

**Pull request merged**